lib/SILOptimizer/Analysis/AccessSummaryAnalysis.cpp

```diff
@@ -0,0 +1,329 @@
+                                             SILFunctionArgument *parameter,
+                                             ParameterSummary &summary,
+                                             FunctionOrder &order) {
+  unsigned parameterIndex = parameter->getIndex();
+
+  // Use a worklist to track parameter uses to be processed.
+  llvm::SmallVector<Operand *, 32> worklist;
+
+  // Start by adding the immediate uses of the parameter to the worklist.
+  worklist.append(parameter->use_begin(), parameter->use_end());
+
+  // Iterate to follow uses of the parameters.
+  while (!worklist.empty()) {
+    Operand *operand = worklist.pop_back_val();
+    SILInstruction *user = operand->getUser();
+
+    switch (user->getKind()) {
+    case ValueKind::BeginAccessInst: {
+      auto *BAI = cast<BeginAccessInst>(user);
+      summary.mergeWith(BAI->getAccessKind(), BAI->getLoc());
+      // We don't add the users of the begin_access to the worklist because
+      // even if these users eventually begin an access to the address
+      // or a projection from it, that access can't begin more exclusive
+      // access than this access -- otherwise it will be diagnosed
+      // elsewhere.
+      break;
+    }
+    case ValueKind::EndUnpairedAccessInst:
+      // Don't diagnose unpaired access statically.
+      assert(cast<EndUnpairedAccessInst>(user)->getEnforcement() ==
+             SILAccessEnforcement::Dynamic);
+      break;
+    case ValueKind::StructElementAddrInst:
+    case ValueKind::TupleElementAddrInst:
+      // Eventually we'll summarize individual struct elements separately.
+      // For now an access to a part of the struct is treated as an access
+      // to the whole struct.
+      worklist.append(user->use_begin(), user->use_end());
+      break;
+    case ValueKind::DebugValueAddrInst:
+    case ValueKind::AddressToPointerInst:
+      // Ignore these uses, they don't affect formal accesses.
+      break;
+    case ValueKind::PartialApplyInst:
+      processPartialApply(info, parameterIndex, cast<PartialApplyInst>(user),
+                          operand, order);
+      break;
+    case ValueKind::ApplyInst:
+      processFullApply(info, parameterIndex, cast<ApplyInst>(user), operand,
+                       order);
+      break;
+    case ValueKind::TryApplyInst:
+      processFullApply(info, parameterIndex, cast<TryApplyInst>(user), operand,
+                       order);
+      break;
+    case ValueKind::CopyAddrInst:
+    case ValueKind::ExistentialMetatypeInst:
+    case ValueKind::LoadInst:
+    case ValueKind::OpenExistentialAddrInst:
+    case ValueKind::ProjectBlockStorageInst:
+      // These likely represent scenarios in which we're not generating
+      // begin access markers. Ignore these for now. But we really should
+      // add SIL verification to ensure all loads and stores have associated
+      // access markers.
+      break;
+    default:
+      llvm_unreachable("Unrecognized argument use");
```

Comment:
Unreachable makes sense since enforcement isn't optional. However, a
hard assumption about the set of SIL operations that can consume an
address needs to be codified in the verifier. I made a note of this,
but I think at least a TODO would be nice.